### PR TITLE
[Diffusion] Add CLIP / CLIP-IQA / ImageReward quality checks for imag…

### PR DESCRIPTION
**Pull request by murthy-l-habana**
## Motivation

The diffusion server tests have no accuracy gate on image output other than an
inline CLIP guard in the XPU FLUX.2-dev case. That guard has two problems:

1. It approximates cosine similarity by dividing CLIP's `logits_per_image` by a
   fixed `100`. `logits_per_image` is `exp(logit_scale) * cosine`, so the divisor
   is an assumption about a *learned* parameter of the checkpoint rather than a
   constant of the metric. It happens to be close for the checkpoint in use, but
   nothing enforces that, and it changes silently under a different or fine-tuned
   CLIP.
2. More importantly, it catches every exception and returns `None`, which
   **silently disables the assertion** whenever setup or inference fails — a
   missing dependency, a failed download, an OOM during scoring. The only
   accuracy check in the case can stop checking anything without the test going
   red.

This PR replaces that guard with three named metrics computed from explicit
cosine similarity, makes an unavailable-but-enforced metric a failure, and ships
an offline sweep tool so a floor is measured before it is enforced.

## Modifications

- **`multimodal_gen/test/quality_metrics.py`** (new) — `compute_clip_score` /
  `compute_clip_iqa` / `compute_image_reward` (plus a batched
  `compute_image_rewards`), `QualityThresholds` with `load_quality_thresholds` /
  `check_quality`, and `paired_delta`.
  - `clip_score` normalises the projected image and text features and takes their
    cosine directly, then multiplies by `100` purely as the reporting convention
    used by `torchmetrics.CLIPScore` and the published FLUX/SD numbers. The
    checkpoint's `logit_scale` is never used, so no fixed value is assumed.
  - It is reimplemented on plain `transformers` rather than calling
    `torchmetrics.multimodal.CLIPScore`, which does
    `get_image_features(...).norm(...)` and raises `AttributeError` on the pinned
    `transformers==5.12.1` where that call returns an output object
    (Lightning-AI/torchmetrics#3244). Backbone is `openai/clip-vit-base-patch32`,
    torchmetrics' default; scores are comparable only within one backbone.
  - Only the metrics a case actually enforces are computed
    (`compute_quality_scores` returns `None` for the rest), and an enforced metric
    that produced no value is reported as a **failure**, never a pass.
- **`multimodal_gen/test/server/quality_thresholds.json`** (new) — per-case
  floors, opt-in through `DiffusionTestCase.run_quality_check` (default `False`).
  A case with the check enabled and no entry in the table fails immediately.
- **`multimodal_gen/test/server/test_server_common.py`** — `_score_quality` /
  `_validate_quality`, wired through the existing `run_case_check("quality", ...)`
  path so a violation reports like every other case check.
- **`multimodal_gen/test/server/testcase_configs.py`** —
  `run_quality_check: bool = False` on `DiffusionTestCase`.
- **`multimodal_gen/benchmarks/bench_image_quality.py`** (new) — scores a
  `--save-output-dir` from `bench_offline_throughput` against the request manifest
  that generated it. Writes `<prefix>_per_image.jsonl` and `<prefix>_summary.json`;
  `--baseline` adds a paired per-prompt delta against another arm, which is what
  makes these metrics usable at small n; `--case-id` checks a whole directory
  against a case's floors; `--limit` truncates after pairing.
- **`multimodal_gen/test/unit/test_quality_metrics.py`** (new) — see
  **Test coverage** below. No GPU and no network.
- **`test/registered/xpu/test_xpu_flux2_dev.py`** — deletes the file's private
  CLIP helper and sets `run_quality_check=True` on `flux2_dev_image_t2i`. **This
  is the only case in the tree that enables the check.**
- **`python/sglang/srt/environ.py`** — `SGLANG_TEST_IMAGE_REWARD_PYTHON`
  (`EnvStr(None)`), in the test section.
- **`python/pyproject.toml`** — a `diffusion-eval` optional extra
  (`piq>=0.8.0`, `torchmetrics>=1.9.0`).
- **`docs/docs/sglang-diffusion/quality_eval.mdx`** + `docs/docs.json` nav entry.

## Exactly what is enforced today

| case | `run_quality_check` | populated floors | enforced when the case runs |
|---|---|---|---|
| `flux2_dev_image_t2i` (`test/registered/xpu/test_xpu_flux2_dev.py`) | **`True`** | `min_clip_score: 20.0` | `clip_score` only |
| `flux_2_klein_image_t2i` | not set (default `False`) | `min_clip_score: 15.0`, `min_image_reward: -1.3` | nothing — the entry is data for `bench_image_quality --case-id` |
| every other case | default `False` | none | nothing |

`min_clip_iqa` is populated for **no** case. So in CI this PR enforces exactly one
threshold: `clip_score >= 20.0` on the FLUX.2-dev XPU case, which is the same gate
the deleted helper intended (it asserted raw cosine `>= 0.20`), carried onto the
`100 * cosine` scale so the case is not silently weakened.

The `flux_2_klein_image_t2i` row is deliberately not wired to a running case: it
exists so `bench_image_quality --case-id flux_2_klein_image_t2i` can check a whole
sweep against measured floors offline. If a maintainer wants it gated in CI, that
is a one-line `run_quality_check=True` plus the ImageReward provisioning below.

## Why `clip_iqa` is unenforced and `clip_score` is only a tripwire

Scoring real output against a black frame and uniform noise (the control images
were produced by a throwaway script, not shipped, then scored with
`bench_image_quality.py`):

| metric | worst healthy | best broken | separation |
|---|---|---|---|
| `clip_score` | 24.70 | 25.28 (black frame) | **-0.58, overlaps** |
| `clip_iqa` | 0.430 | 0.773 (uniform noise) | **-0.34, overlaps** |
| `image_reward` | -0.587 | -2.041 (black frame) | **+1.45** |

A black frame out-scores the worst healthy image on CLIP, and uniform noise lands
inside the healthy bulk on CLIP-IQA. So neither can separate a broken image from a
good one at any floor: `min_clip_iqa` is therefore populated nowhere, and the
`min_clip_score` floors that do exist sit far below the healthy distribution and
are documented as **decode-failure tripwires** — they catch "the pipeline returned
something that is not an image of the prompt", not a quality regression.
`image_reward` is the only one of the three whose distributions do not overlap, so
it is the only one a single-image floor can be based on; that is why the one
measured `min_image_reward` exists, on the case that has a measured entry.

For anything finer than a tripwire, use `--baseline`: a paired per-prompt delta
between two arms resolves differences far below what any floor can, because the
per-prompt spread of all three metrics dominates the arm-to-arm difference.

## Dependencies and how the second interpreter is provisioned

| metric | needs | why |
|---|---|---|
| `clip_score` | nothing beyond sglang's pinned `transformers` | computed in-tree |
| `clip_iqa` | `pip install "sglang[diffusion-eval]"` (`torchmetrics>=1.9.0`, `piq>=0.8.0`) | `piq` holds the canonical CLIP-IQA weights, which torchmetrics refuses to load without it |
| `image_reward` | a **second interpreter** named by `SGLANG_TEST_IMAGE_REWARD_PYTHON` | ImageReward imports `ImageReward/ReFL.py`, which requires `transformers<5`, while sglang pins `transformers==5.12.1`. The two cannot share an environment |

```bash
python -m venv /opt/imagereward
/opt/imagereward/bin/pip install image-reward 'transformers==4.46.3' 'diffusers==0.31.0'
export SGLANG_TEST_IMAGE_REWARD_PYTHON=/opt/imagereward/bin/python
```

sglang never imports ImageReward. `quality_metrics.py` runs a small stdin/stdout
script under that interpreter, handing images over by path, one launch per sweep
(the checkpoint is ~1.7 GB, so per-image launches would dominate).

**CI needs no change for this PR.** The only case with the check enabled enforces
`clip_score`, which needs neither the extra nor the second interpreter. The
contract if that changes is deliberate and worth stating plainly: a case that
enforces `min_image_reward` **fails** when `SGLANG_TEST_IMAGE_REWARD_PYTHON` is
unset — it does not skip. So gating a CI case on `image_reward` requires the runner
image to provision that venv first, and reviewers should treat the env var as a
prerequisite of enabling such a floor, not as an optional enhancement. If a
skip-when-absent policy is preferred for CI, say so and I will add it as an
explicit mode rather than as silent behaviour.

## Scope of the single threshold table

The table is not split per device, and that is a claim about **these floors**, not
about score invariance. Output quality does vary with precision, attention and
kernel selection, scheduler, seed and dependency versions, not merely GPU family.
The floors here are set far enough below the healthy distribution that they survive
that variation — which is exactly what makes them tripwires rather than quality
gates — whereas `perf_baselines/` has to be per-device because latency has no such
margin. If some configuration genuinely scores near a floor, the right fix is a
case-specific entry (or `--baseline` comparison), not a per-device table.

## Test

```bash
# 1. unit tests -- no GPU, no network, no model downloads
pytest python/sglang/multimodal_gen/test/unit/test_quality_metrics.py -v

# 2. the gated registered case -- COLLECTION ONLY here (see Limitations)
pytest test/registered/xpu/test_xpu_flux2_dev.py --collect-only -q

# 3. offline sweep, end to end: generate, then score
python -m sglang.multimodal_gen.benchmarks.bench_offline_throughput \
    --model-path black-forest-labs/FLUX.2-klein-9B \
    --request-manifest prompts.jsonl \
    --save-output-dir out_arm1/ \
    --height 1024 --width 1024 \
    --num-inference-steps 4 --guidance-scale 1.0 --seed 42 \
    --batch-size 1

python -m sglang.multimodal_gen.benchmarks.bench_image_quality \
    --images out_arm1/ \
    --request-manifest prompts.jsonl \
    --out-prefix arm1 \
    --metrics clip_score clip_iqa image_reward \
    --case-id flux_2_klein_image_t2i \
    --log-every 25

# 4. paired comparison of a second arm against the first, and a truncated pass
python -m sglang.multimodal_gen.benchmarks.bench_image_quality \
    --images out_arm2/ \
    --request-manifest prompts.jsonl \
    --out-prefix arm2 \
    --metrics clip_score clip_iqa image_reward \
    --baseline arm1_per_image.jsonl \
    --limit 100 --log-every 25
```

Generate on one device: `bench_offline_throughput` names outputs
`<index-within-shard>-<request_id>`, and `bench_image_quality` re-checks that index
and id against the manifest row, so a sharded run raises `PAIRING ERROR` on every
shard after the first. That check is deliberate — mislabelling prompts would
silently corrupt every score.

### Test coverage

Behaviours asserted by the unit tests, all without a GPU, network access or model
weights (`transformers`, `torchmetrics` and `subprocess` are faked):

- an enforced metric that produced no value fails, rather than passing — the
  regression this PR exists to prevent;
- an unenforced metric is neither computed nor required;
- threshold lookup merges `defaults`, and an unknown case id yields no floors;
- the shipped table loads and enforces at least one floor;
- CLIP-IQA's `data_range` is pinned to the scale of the tensor actually passed;
- `clip_score` reads projected features off the `transformers` 5 output object;
- ImageReward scores are parsed from the subprocess's last JSON line, tolerating
  loader noise on stdout;
- a sweep launches the ImageReward interpreter exactly once for a whole directory;
- ImageReward with no second interpreter configured raises, rather than skipping;
- `paired_delta` returns the mean per-image difference and its own standard error;
- pairing rejects images whose index or request id does not match the manifest;
- every case with `run_quality_check=True` has an entry in the thresholds table
  (a repo-consistency test, so the two cannot drift).

### Limitations

- **`test/registered/xpu/test_xpu_flux2_dev.py` has been collected, not executed
  end to end** (`1 test collected`) — the environment used had neither the
  FLUX.2-dev weights nor the disk for them, and the case is registered to a 4-GPU
  nightly XPU suite. The server-side wiring is therefore covered by the unit tests
  plus collection, and the metrics themselves by the offline sweep. A CI run of
  that case is the missing piece, and reviewers should treat it as unverified until
  it goes green.
- The floors are measured for FLUX.2-dev and FLUX.2-klein only. Other image cases
  can opt in as their entries are measured; the repo-consistency test makes
  enabling the flag without an entry a failure rather than a silent pass.

## Checklist

- [x] Formatted with the repo's pre-commit configuration (black, isort, ruff
      `F401,F821,UP037`).
- [x] Unit tests added, runnable without a GPU, network or model weights.
- [x] Documentation added (`docs/docs/sglang-diffusion/quality_eval.mdx`) and
      registered in `docs/docs.json`.
- [x] New env var follows `.claude/skills/env-var-conventions` — typed descriptor
      in `Envs`, read only through `envs.X.get()`, overridden in tests with
      `.override(...)`.

---
### CI States

Latest PR Test (Base): :x: [Run #35187383637](https://github.com/sgl-project/sglang/actions/runs/35187383637)
Latest PR Test (Extra): :x: [Run #35187383277](https://github.com/sgl-project/sglang/actions/runs/35187383277)
Latest PR Test (AMD ROCm 10): :x: [Run #35187383646](https://github.com/sgl-project/sglang/actions/runs/35187383646)

### CODE REVIEW

File: python/sglang/multimodal_gen/test/server/quality_thresholds.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    "defaults": {},
+    "cases": {
+        "flux2_dev_image_t2i": {
+            "_comment": "Carried over from the previous inline XPU guard, which asserted raw cosine >= 0.20; same gate on the 100*cosine scale, kept so this case is not silently weakened. Read it as a smoke floor on one fixed prompt, not a quality gate: a black frame scores 17.2-25.3, which is inside the range healthy images occupy. clip_iqa/image_reward are unmeasured for FLUX.2-dev, so they stay unenforced.",
+            "min_clip_score": 20.0
```
Comment (msinnha1) [inline python/sglang/multimodal_gen/test/server/quality_thresholds.json]: this seems contradicting with the finding of black frame clip_score of 17.2-25.3

Comment (murthy-l-habana) [reply python/sglang/multimodal_gen/test/server/quality_thresholds.json]: Addressed it, made it 25 and test is passing for 100 images.
